pkg/controllers/node/ipam.go

```diff
@@ -540,27 +642,30 @@ func (c *ipamController) cleanup() error {
 	if storedErr != nil {
 		return storedErr
 	}
+
+	log.Debug("IPAM sync completed")
 	return nil
 }
 
 // garbageCollectIPs checks all known allocations and GCs any confirmed leaks.
 func (c *ipamController) garbageCollectIPs() error {
-	// TODO: Do this without iterating every allocation. As we mark them, we can index known leaks
-	// for quicker lookups.
-	for _, allocations := range c.allocationsByNode {
-		for _, a := range allocations {
-			if a.isConfirmedLeak() {
-				logc := log.WithFields(a.fields())
-				logc.Info("Garbage collecting leaked IP address")
-				if err := c.client.IPAM().ReleaseByHandle(context.TODO(), a.handle); err != nil {
-					if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
-						logc.WithField("handle", a.handle).Debug("IP already released")
-						continue
-					}
-					logc.WithError(err).WithField("handle", a.handle).Warning("Failed to release leaked IP")
-					return err
+	for handle, a := range c.confirmedLeaks {
+		if a.isConfirmedLeak() {
+			logc := log.WithFields(a.fields())
+			logc.Info("Garbage collecting leaked IP address")
+			if err := c.client.IPAM().ReleaseByHandle(context.TODO(), a.handle); err != nil {
```

Comment:
Have we ever seen allocations get leaked with a missing handle resource?  Just wondering if we needed to deal with that case (in a future increment).  I suppose we could do that by recreating the handle resource from the metadata in the IPAM block(s); or we could pass in a hint to say "I'm expecting the handle to match this IP" here.

Reply:
I seem to recall us hitting a strange issue like this once before. Definitely something we should try to handle in the future - along with other "inconsistent datastore" type situations.